### PR TITLE
#811 number records removed from sync 

### LINKED ISSUE #811: Number records removed from sync
[A new KDD on mSupply Server](https://github.com/sussol/msupply/pull/11767/files). Pasted below in case one cannot access:

<details>
# KDD-0012: Sync v5 number sequences not to sync

- _Date_: 2022-11-15
- _Deciders_: @Chris-Petty @andreievg
- _Status_: Decided
- _Outcome_: Don't sync number records. Generate them on use, and clean up when stores move around.

## Context

mSupply uses tables `number` and `number_reuse` to manage serials numbers on invoices (oms shipments), requisitions and several other things. These number tables have historically not been included in sync which has lead to some headaches with serial numbers seemingly resetting as stores move from one site to another, or sites get devices replaced and initialised.

Recently in sync v5 it was decided to sync them to resolve this problem. This has been implemented in mSupply desktop (4D) but in testing and addressing number sequence sync in omSupply it has become apparent it will not work:

- When sites are upgraded to omSupply, they may not have been previously upgraded to the latest version of mSupply desktop that syncs number sequences. This results in omSupply not getting them and generating its own that start from `1`.
- Desktop tracks requisition sequences with `requisition_number_for_store_`.
- Mobile `requisition_serial_number_for_store_`.
- Omsupply has split the sequences into `request_requisition_for_store_` and `response_requisition_for_store_` for request and response requisitions respectively.

This causes a mess. If all these applications sync their number sequences, when a store is moved to another site or the site is upgraded from one app to another, the sequences don't get reused because the keys don't match. Omsupply has diverged (in a good way) from what desktop was doing, so it's not immediately clear how to translate this. We actually thought that mobile was doing separate sequences too but it appears in code that it doesn't! It has a second sequence for another field `requisition.requester_reference` that seems to just be redundant.

We need a solution that:

- Allows stores to move from one site to another using potentially a different app and continue allocate serial numbers correctly (including number reuse)
- Allows a site to initialised in omSupply that was previously used on an older version of desktop or mobile and continue allocate serial numbers correctly (including number reuse)
- Regarding number_reuse, after a discussion with @craigdrown we've agreed that it's not used much. In operation if a store moves from one site to another, a site changes app or the number reuse pref is turned on, then we **do not** need to backfill the `number_reuse` table. Just continue from the current highest serial number.

### Small caveat on number reuse

omSupply uses 2 sequences for the 2 types of requisition unlike the old apps. While we don't currently support number reuse for requisitions, it could lead to some odd behaviours if we did:

- OmSupply store moves to desktop or mobile - We accept that it'll appear that there are duplicate sequences across the 2 requisition types.
- Desktop/mobile store moves omSupply - We accept that both requisition types will appear to have gaps in the sequences. This may wreck havoc with `number_reuse`, though! If the pref is on the intuition will be to reuse all the gaps which will seem like an odd behaviour.

### Option 1

The central server on initialisation sorts out making sure the correct sequences exist.

As described in the context above, we can rule this one out as it is flawed:

- Because each app has a different sequence key for requisition sequences, we have to make the central server client aware to make sure we're sending out `number` records with the correct key for the client app 😭
- Stores can be moved from site to site outside of initialisation. So solutions should address this too. The central server would also have to make sure it does it correctly when moving a store making sure to satisfy the above point.

### Option 2

The apps generate number sequences on initialisation and whenever a store is synced and set to active. Don't bother syncing number sequences.

- The remote sites would have to be **certain** that they have synced **all** all the records relating to the sequences before generating the number records, or face gaps and/or duplicates.
- The central server would have to make sure that for stores that are made active on itself, it generates the sequence for use as a remote site would.

_Pros:_

- Each remote app can maintain its own style of serial numbering
- At the moment the only tenable approach

_Cons:_

- Have to trust the remote apps to handle it correctly - this is fine, we already trust them to do serials numbering correctly regardless
- Have to implement it in every app. Not too hard
- Have to undo some existing work. Always the way 😉

### Option 3

When ever an app requests the next number in the sequence, it checks for the `number` record. If it exists, use it. If not, get the `max(invoice_num)+1` and save the `number` record.
When ever a store becomes active on a site, it should delete the `number` and `number_reuse` records for the store so that it may reset as above. On becoming inactive would be equivalent.

_Pros:_

- Super easy and effective

_Cons:_

- Less fun? Seems pretty solid really.

## Decision

Option 3 (for all apps).

Easiest approach, with least complexity that is quite robust. Code largely in one place. Not great for `number_reuse` though.

- Option 1 is not functional
- Option 2 works, but is complex and scattered.

Did consider option 2 for Desktop/4D and option 3 for omSupply, but after clarifying number reuse behaviours option 3 is adequate for mSupply desktop and way easier than option 2.

## Consequences

- mSupply desktop, open mSupply needs to implement option 3.
  - Whenever an app requests the next number in the sequence, it checks for the `number` record. If it exists, use it. If not, get the `max(invoice_num)+1` and save the `number` record.
  - Whenever a store becomes active on a site, it should delete the `number` and `number_reuse` records for the store so that it may reset as above. On becoming inactive would be equivalent.
  - Remove any code for syncing the `number` and `number_reuse` tables.
  - Migration for removing any `number` and `number_reuse` records for stores not active on the datafile/site.
</details>

To handle numbers sequences correctly:

- [x] Whenever getting the next number in the sequence, it checks for the `number` record. If it exists, use it. If not, get the `max(invoice_num)+1` and save the `number` record.
- [ ] Whenever a store becomes active on a site, it should delete the `number` and `number_reuse` records for the store so that it may reset as above. On becoming inactive would be equivalent.
- [x] Remove any code for syncing the `number` and `number_reuse` tables.
- [ ] If receiving `number` or `number_reuse` records in sync, just ignore them.
- [ ] Migration for removing any `number` and `number_reuse` records for stores not active on the datafile/site.

### PR DESCRIPTION
**Pull request by roxy-dao**
Closes #811 

- Check for type numbers if does not exist and increment those by 1 to save into the number table.
- Remove all number syncing. 
- Delete any existing number entries if the store has been moved out of the site

### CODE REVIEW

File: server/service/src/number.rs [hunk 1]
```diff
@@ -8,9 +11,74 @@ pub fn next_number(
...
     let next_number = connection.transaction_sync(|connection_tx| {
         let repo = NumberRowRepository::new(&connection_tx);
-        let next_number = repo.get_next_number_for_type_and_store(r#type, store_id)?;
-
-        Ok(next_number.number)
+        let number = repo.find_one_by_type_and_store(r#type, store_id)?;
+
+        if number.is_some() {
+            let next_number = repo.get_next_number_for_type_and_store(r#type, store_id, None)?;
+            Ok(next_number.number)
```
Comment (andreievg) [inline server/service/src/number.rs:18]: :nit early return

File: server/service/src/number.rs [hunk 2]
```diff
@@ -8,9 +11,74 @@ pub fn next_number(
...
+        } else {
+            match r#type {
+                NumberRowType::InboundShipment => {
+                    let find_number = InvoiceRowRepository::new(&connection_tx)
+                        .find_max_invoice_number(InvoiceRowType::InboundShipment, store_id)?
+                        .map(|n| n + 1);
+                    let next_number =
```
Comment (andreievg) [inline server/service/src/number.rs:25]: I think we should return `found_number` from `match` (do early return for program only). And then just have one `let next_number = repo.get_next_number_for_type_and_store(r#type, store_id, `found_number`)?;`

File: server/service/src/sync/mod.rs [hunk 3]
```diff
@@ -79,3 +80,31 @@ impl ActiveStoresOnSite {
...
+
+#[derive(Error, Debug)]
+pub(crate) enum GetInactiveStoresOnSiteError {
+    #[error("Database error while getting inactive store on site")]
+    DatabaseError(RepositoryError),
+    #[error("Site id is not set in database")]
+    SiteIdNotSet,
+}
+
+pub(crate) fn get_inactive_store_ids(
```
Comment (andreievg) [inline server/service/src/sync/mod.rs:92]: I think this logic should be in another PR/branch, to be done when we implement store switch (it needs extensive check and having it here now would mean it can potentially have a use case we haven't thought about/tested)

**Review by andreievg — CHANGES_REQUESTED** [review]
Thanks, look good, a couple of small requests (instead of removing and moving delete inactive, can just comment out for now, of course open to discussion).

File: server/service/src/sync/mod.rs (hunk 3 above)
Comment (Chris-Petty) [reply server/service/src/sync/mod.rs:92]: Sorry, I had already made a [second PR](https://github.com/openmsupply/open-msupply/pull/943) with just the other changes! I'll rebase this one to stack it correctly.

Comment (Chris-Petty) [reply server/service/src/sync/mod.rs:92]: Done. Weird, the diff isn't picking up that the new base PR has most of the commits in this PR...

File: server/service/src/number.rs (hunk 1 above)
Comment (Chris-Petty) [reply server/service/src/number.rs:18]: Not entirely sure what the request is here, is it for `return Ok(next_number.number);`? I'll go for it with the other PR - it's consistent with the match statement which uses `return` syntax. FWIW it's returning evaluated value into another variable (that it shadows):

image

Comment (andreievg) [reply server/service/src/number.rs:18]: I was meaning something like this: https://github.com/openmsupply/open-msupply/compare/%23811-Number-records-removed-from-sync-...811-refactor-example

**Comment by andreievg** [comment]
@Chris-Petty should this be based off develop ?

File: server/service/src/number.rs (hunk 1 above)
Comment (Chris-Petty) [reply server/service/src/number.rs:18]: Oh heck yea that's way shorter! Thanks

**Comment by DhanyaHerath** [comment]
@mark-prins do you think this isn't required for an stable android deployment?

**Comment by Chris-Petty** [comment]
> @Chris-Petty should this be based off develop ?

No this PR is meant to just have the code for deleting number records after stores' site IDs change. But it's not picking up the cherry picked commits in the new base are there. I'm going to close this PR for discussion on that part. Please review https://github.com/openmsupply/open-msupply/pull/943 instead.

**Comment by Chris-Petty** [comment]
> @mark-prins do you think this isn't required for a stable android deployment? (just noticed you removed the milestone)

Appears that he's just removed the _PR_ from the milestone, not the issue. Bot automatically setting it for PRs.
